assertsprocessor/span_enrichment.go

```diff
@@ -15,67 +17,108 @@ const (
 	AssertsRequestTypeInternal     = "internal"
 )
 
-type ErrorTypeConfig struct {
-	ValueExpr string `mapstructure:"value_match_regex"`
-	ErrorType string `mapstructure:"error_type"`
-}
-
-func (eTC *ErrorTypeConfig) compile() (*errorTypeCompiledConfig, error) {
-	compile, err := regexp.Compile(eTC.ValueExpr)
-	if err != nil {
-		return nil, err
-	} else {
-		e := &errorTypeCompiledConfig{
-			errorType:    eTC.ErrorType,
-			valueMatcher: compile,
-		}
-		return e, nil
-	}
-}
-
-type errorTypeCompiledConfig struct {
-	valueMatcher *regexp.Regexp
-	errorType    string
-}
-
 type spanEnrichmentProcessor interface {
 	enrichSpan(namespace string, service string, span *ptrace.Span)
 }
 
 type spanEnrichmentProcessorImpl struct {
 	logger           *zap.Logger
-	errorTypeConfigs map[string][]*errorTypeCompiledConfig
-	requestBuilder   requestContextBuilder
+	customAttributes map[string]map[string][]*customAttributeConfigCompiled
+	configRWMutex    sync.RWMutex
 }
 
-func buildEnrichmentProcessor(logger *zap.Logger, config *Config, requestBuilder requestContextBuilder) *spanEnrichmentProcessorImpl {
-	processor := spanEnrichmentProcessorImpl{
-		logger:           logger,
-		errorTypeConfigs: map[string][]*errorTypeCompiledConfig{},
-		requestBuilder:   requestBuilder,
+func buildEnrichmentProcessor(logger *zap.Logger, config *Config) (*spanEnrichmentProcessorImpl, error) {
+	compiledAttributes, err := buildCompiledConfig(logger, config)
+	if err == nil {
+		processor := spanEnrichmentProcessorImpl{
+			logger:           logger,
+			customAttributes: compiledAttributes,
+		}
+		return &processor, err
+	} else {
+		return nil, err
 	}
-	for attrName, errorConfigs := range config.ErrorTypeConfigs {
-		logger.Debug("Compiling error type configs for", zap.String("attribute", attrName))
-		processor.errorTypeConfigs[attrName] = make([]*errorTypeCompiledConfig, 0)
-		for _, errorConfig := range errorConfigs {
-			compile, _ := regexp.Compile(errorConfig.ValueExpr)
-			processor.errorTypeConfigs[attrName] = append(processor.errorTypeConfigs[attrName],
-				&errorTypeCompiledConfig{
-					errorType:    errorConfig.ErrorType,
-					valueMatcher: compile,
-				})
-			logger.Debug("Compiled and added error type config",
-				zap.String("error type", errorConfig.ErrorType),
-				zap.String("attr value regexp", errorConfig.ValueExpr))
+}
+
+func buildCompiledConfig(logger *zap.Logger, config *Config) (map[string]map[string][]*customAttributeConfigCompiled, error) {
+	compiledAttributes := map[string]map[string][]*customAttributeConfigCompiled{}
+	for targetAtt, attrConfigsByServiceKey := range config.CustomAttributeConfigs {
+		compiledAttributes[targetAtt] = map[string][]*customAttributeConfigCompiled{}
+		for serviceKey, attrConfigs := range attrConfigsByServiceKey {
+			if compiledAttributes[targetAtt][serviceKey] == nil {
```

Comment:
This if check can be removed as the list is being iterated in the for loop below this